File: 08. Trees/14. Bottom View/main.cpp
#include <bits/stdc++.h>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// BFS Approach for Bottom View
vector<int> bottomViewBFS(TreeNode* root) {
    if (!root) return {};

    map<int, int> bottomNodes;
    queue<pair<TreeNode*, int>> q;
    q.push({root, 0});

    while (!q.empty()) {
        auto [node, hd] = q.front();
        q.pop();

        bottomNodes[hd] = node->val;  // Store last encountered node at this HD

        if (node->left) q.push({node->left, hd - 1});
        if (node->right) q.push({node->right, hd + 1});
    }

    vector<int> result;
    for (auto& [_, val] : bottomNodes) {
        result.push_back(val);
    }
    return result;
}

// DFS Approach for Bottom View
void dfs(TreeNode* node, int hd, int depth, map<int, pair<int, int>>& bottomNodes) {
    if (!node) return;

    // If depth is greater, update bottom view
    if (!bottomNodes.count(hd) || depth >= bottomNodes[hd].second) {
        bottomNodes[hd] = {node->val, depth};
    }

    // Traverse left and right subtrees
    dfs(node->left, hd - 1, depth + 1, bottomNodes);
    dfs(node->right, hd + 1, depth + 1, bottomNodes);
}

vector<int> bottomViewDFS(TreeNode* root) {
    if (!root) return {};

    map<int, pair<int, int>> bottomNodes;
    dfs(root, 0, 0, bottomNodes);

    vector<int> result;
    for (auto& [_, val] : bottomNodes) {
        result.push_back(val.first);
    }
    return result;
}

// Main function
int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->left->right->right = new TreeNode(5);
    root->left->right->right->right = new TreeNode(6);

    vector<int> resultBFS = bottomViewBFS(root);
    vector<int> resultDFS = bottomViewDFS(root);

    cout << "Bottom View (BFS): ";
    for (int val : resultBFS) cout << val << " ";
    cout << endl;

    cout << "Bottom View (DFS): ";
    for (int val : resultDFS) cout << val << " ";
    cout << endl;

    return 0;
}

File: 08. Trees/14. Bottom View/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// BFS Approach for Bottom View
function bottomViewBFS(root) {
    if (!root) return [];

    let bottomNodes = new Map();
    let queue = [[root, 0]];

    while (queue.length > 0) {
        let [node, hd] = queue.shift();
        bottomNodes.set(hd, node.val);  // Store last encountered node at this HD

        if (node.left) queue.push([node.left, hd - 1]);
        if (node.right) queue.push([node.right, hd + 1]);
    }

    return [...new Map([...bottomNodes.entries()].sort())].map(([_, val]) => val);
}

// DFS Approach for Bottom View
function dfs(node, hd, depth, bottomNodes) {
    if (!node) return;

    if (!bottomNodes.has(hd) || depth >= bottomNodes.get(hd)[1]) {
        bottomNodes.set(hd, [node.val, depth]);
    }

    dfs(node.left, hd - 1, depth + 1, bottomNodes);
    dfs(node.right, hd + 1, depth + 1, bottomNodes);
}

function bottomViewDFS(root) {
    if (!root) return [];

    let bottomNodes = new Map();
    dfs(root, 0, 0, bottomNodes);

    return [...new Map([...bottomNodes.entries()].sort())].map(([_, val]) => val[0]);
}

function main() {
    let root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.right = new TreeNode(4);
    root.left.right.right = new TreeNode(5);
    root.left.right.right.right = new TreeNode(6);

    console.log("Bottom View (BFS):", bottomViewBFS(root));
    console.log("Bottom View (DFS):", bottomViewDFS(root));
}

main();
